src/mirrors.rs

```diff
@@ -0,0 +1,118 @@
+/// Implementation of the PostgreSQL server (database) protocol.
+/// Here we are pretending to the a Postgres client.
+use bytes::{BytesMut};
+
+use parking_lot::Mutex;
+use tokio::sync::mpsc::{Receiver, Sender, channel};
+
+use crate::config::{Address, User};
+use crate::pool::ClientServerMap;
+use crate::server::Server;
+use crate::stats::get_reporter;
+
+pub enum MirrorOperation {
+    Send(BytesMut),
+    Receive
+}
+pub struct MirrorUnit {
+    pub address: Address,
+    pub user: User,
+    pub database: String,
+    pub bytes_rx: Receiver<MirrorOperation>,
+    pub exit_rx: Receiver<()>,
+}
+
+impl MirrorUnit {
+    pub fn begin(mut self, server_id: i32) {
+        tokio::spawn(async move {
+            let mut server = Server::startup(
+                server_id,
+                &self.address.clone(),
+                &self.user.clone(),
+                self.database.as_str(),
+                ClientServerMap::default(),
+                get_reporter()
+            ).await.unwrap();
+
+            loop {
+                tokio::select! {
+                    _ = self.exit_rx.recv() => {
+                        break;
+                    }
+                    op = self.bytes_rx.recv() => {
+                        match op {
+                            Some(MirrorOperation::Send(bytes)) => {
+                                server.send(&bytes).await.unwrap();
+                            }
+                            Some(MirrorOperation::Receive) => {
+                                server.recv().await.unwrap();
+                            }
+                            None => {
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        });
+    }
+}
+pub struct MirroringManager {
+    pub byte_senders: Vec<Sender<MirrorOperation>>,
+    pub exit_senders: Vec<Sender<()>>
+}
+impl MirroringManager {
+    pub fn from_addresses(user: User, database: String, addresses: Vec<Address>) -> MirroringManager {
+        let mut byte_senders : Vec<Sender<MirrorOperation>> = vec![];
+        let mut exit_senders : Vec<Sender<()>> = vec![];
+
+        addresses.iter().for_each(|mirror| {
+            let (bytes_tx, bytes_rx) = channel::<MirrorOperation>(500);
+            let (exit_tx, exit_rx) = channel::<()>(1);
+            let mirror_unit = MirrorUnit {
+                user: user.clone(),
+                database: database.to_owned(),
+                address: mirror.clone(),
+                bytes_rx,
+                exit_rx,
+            };
+            exit_senders.push(exit_tx.clone());
+            byte_senders.push(bytes_tx.clone());
+            mirror_unit.begin(rand::random::<i32>());
+        });
+
+        Self {
+            byte_senders: byte_senders,
+            exit_senders: exit_senders
+        }
+    }
+
+    pub fn send(self: &mut Self, bytes: &BytesMut) {
+        let cpy = bytes.clone();
```

Comment:
This is happening on the client<->server task so I would like to avoid it as much as possible. Not sure how to do it given `server.send` requires `BytesMut`

Reply:
I figured it out. I will change `send` to use `ByteMut.freeze` which yields a frozen chunk of memory that is cheap to clone of the type `Bytes`

Reply:
Well, changing `server.send` to take `Bytes` instead of `BytesMut` will require doing a ton of `BytesMut.clone()` in the server critical path which is not something I want to do.

Instead, I have the communication channel use `Bytes` and I make just one clone per server  to create a frozen `Bytes` object, when I do fan-out I send that same frozen object across the channel and then create a new BytesMut in the mirror task